### PR TITLE
Change the search_analyzer for uri

### LINKED ISSUE #1537: Analyse `uri` field when searching
With openannotation/annotator-store just merging the change to use match queries, we can change the analyzer for the `uri` now to tokenize the field and strip things, potentially the protocol, query string, and fragment for instance. This would introduce some optimism in our annotation results to account for poor practices by webmasters serving equivalent content and various uses of queries and hashes that don't dramatically affect primary content.

### PR DESCRIPTION
**Pull request by gergely-ujvari**
This PR carries the intention to have better matches on uri searches.
It extends both the search and index analyzers with uris stripped from its protocol or query or fragment.

Fix #1537

### CODE REVIEW

File: h/models.py [hunk 1]
```diff
@@ -115,13 +115,25 @@ def __acl__(self):
...
+                    '://(.+)\\?',
+                    '://(.+)\\#'
+                ]
+            },
+            'uri_search': {
+                'type': 'pattern_capture',
+                'preserve_original': '1',
+                'patterns': [
+                    '://(.+)',
+                    '://(.+)\\?',
```
Comment (csillag) [inline h/models.py]: Does this also remove the query string?
Ie. does this make http://site.com/content?id=1 equivalent with  http://site.com/content?id=2 ?
If yes, do we want that?
I don't think so.

Removing the hash part is warranted; removing the query string is much more controversial. 
I would guess that this would give us way more false positives, compared to the cases when it's actually useful.

Comment (gergely-ujvari) [inline h/models.py]: Yes and no at the same times. Yes, this can produce loading orphan annotation but at the same time we should treat
```http://site.com/content?mobile=0``` and ```http://site.com/content?mobile=1``` the same in the sake of annotations.

Comment (csillag) [inline h/models.py]: > Yes, this can produce loading orphan annotation

Orphan annotations are the least of our problems. Real false matches are worse. We can easily find the same piece of text in different documents, talking about different things. And then we would anchor the annotations to place where they don't belong to.

> but at the same time we should treat ```http://site.com/content?mobile=0``` and ```http://site.com/content?mobile=1``` the same in the sake of annotations.

That's why we have the support for `canonical-uri` and friends - right? Those should be linked using some meta tags.

Comment (gergely-ujvari) [inline h/models.py]: Doesn't our various selectors are for making this harder to happen? (I mean we have a selector which saves the text before and after the selected part)

But yes, that can happen, we question is, which is the least problematic.

About canonical-uri: they are not in our control, we cannot force any page owner to use them.

Comment (csillag) [inline h/models.py]: > Doesn't our various selectors are for making this harder to happen? (I mean we have a selector which saves the text before and after the selected part)

Our selectors are OR-es; not AND-ed. In other words, if any of the matches, we attach.
The prefix and suffix we are saving is only 32 characters long. Which means if we have a match within a 80-char sentence, and the same sentence is repeated in an other document, we will have a false match.

> About canonical-uri: they are not in our control, we cannot force any page owner to use them.

We are also not obligated to make H work on broken pages that omit the required meta-info.

  * * *

To summarize: the query parameters typically address *different content*. We should not remove them.

**Comment by gergely-ujvari** [comment]
Okay, I've removed the part which strips the query part.

**Comment by tilgovi** [comment]
LGTM do we want tests?

**Comment by gergely-ujvari** [comment]
I'll write some.

**Comment by gergely-ujvari** [comment]
On second thought: These tests would require us to create an ES test repo and make some operations into it. Currently we have nothing like that. Do we want to write tests like that, like the annotator-store? I'm asking this, because we reach ES via the annotator-store but we set up our custom mapping.

**Comment by gergely-ujvari** [comment]
@nickstenning: A question. You put it back to uncategorized because of the testing framework?

**Comment by nickstenning** [comment]
Don't worry about it -- we just don't need pull requests in the product backlog. You can keep working on it!

**Comment by gergely-ujvari** [comment]
Okay, I've added stub for testing elasticsearch mapping and a test case for the `uri` field.
It is ready for merging / review.

**Comment by gergely-ujvari** [comment]
As soon as I fix travis.

**Comment by gergely-ujvari** [comment]
Travis is fixed too. So really ready for review and merging.

**Comment by nickstenning** [comment]
For the sake of reusability and consistency, I'd rather you configured the elasticsearch connection in a py.test fixture in [conftest.py](https://github.com/hypothesis/h/blob/71b03dff09c72d77fd3b452cda080a072c49eb02/h/conftest.py), with the settings read from `test.ini` much as we do for sqlalchemy. If you do this, you could then remove all the boilerplate from the test file itself. So something like this goes in `conftest.py`:

```
import pytest

from h.api import create_db, delete_db
from h.api import store_from_settings

@pytest.fixture
def es_connection(request, settings):
    es = store_from_settings(settings)

    request.add_finalizer(delete_db)
    create_db()
    es.conn.cluster.health(wait_for_status='yellow')
```

And then you can use the following in your test file:

```
import pytest

@pytest.mark.usefixtures('es_connection')
def test_uri_mapping():
    ...
```

**Comment by gergely-ujvari** [comment]
@nickstenning: Cool :+1:  Thanks for the tip. I've changed the test as you suggested.

File: h/test/models_test.py [hunk 2]
```diff
@@ -1,10 +1,13 @@
 # -*- coding: utf-8 -*-
 import unittest
+import pytest
 
 from pytest import fixture, raises
 from pyramid import security
 
+from annotator import es
```
Comment (nickstenning) [inline h/test/models_test.py]: Unused import.

File: h/test/models_test.py [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
 from h import models
+from h.api import create_db
```
Comment (nickstenning) [inline h/test/models_test.py]: Unused import.

**Comment by nickstenning** [comment]
LGTM. :cake:
